### DISCUSSION TITLE
Has anyone else run into problem with the MPC comet file?

### DISCUSSION THREAD

**Discussion by ursomniac** [skyfielders/python-skyfield#704] [Q&A] [answered]
All of a sudden today, the code I cribbed from the examples for comets is failing on the
```
with load.open('CometEls.txt') as f:
        comets = mpc.load_comets_dataframe(f)
```
code.   I'm getting an error: ParserError: Error tokenizing data. C error: Expected 12 fields in line 488, saw 13
which if line 488 refers to the CometEls.txt file, it just happens to be where the file goes from Recent comets to periodic comets (i.e., line 488 is 1P/Halley).

I'm also seeing warnings from mpc.py:
FutureWarning: In a future version of pandas all arguments of read_fwf except for the argument 'filepath_or_buffer' will be keyword-only.
  df = pd.read_fwf(fobj, colspecs, names=names, dtypes=_MPCORB_DTYPES,

but until today this hadn't been happening.

Is anything seeing this happen, and if so, could you provide insight?

**Comment by jurezakrajsek** [comment]
Hi

The MPC comet file seems to be corrupted from time to time. It takes a few days before they correct the issues.
I have run into problems with the MPC files in the past.

I found another reliable source for the orbital elements, which are updated regularly:
http://astro.vanbuitenen.nl/cometelements?format=mpc&mag=obs

For some comets I make a direct query to MPC ephemeris service to obtain the orbital element line. 
You could use a script like this:

```
import urllib.parse
import urllib.request


def get_comet_mpc_orbit(comet_name):
    orb = None

    if not orb:
        comet_name = comet_name.split('(')[0].strip()
        response = get_mpc_ephemeris_response(-1, comet_name)
        result = response.readlines()
        try:
            orb = result[0]
            orb = orb.decode('UTF-8')
        except Exception as e:
            orb = None

    return orb or None


def get_mpc_ephemeris_response(field, comet_name):
    # A CGI query is produced
    query = {
        'ty': 'e', 'TextArea': comet_name, 'd': '', 'l': '', 'i': '', 'u': 'd', 'uto': '0', 'c': '106', 'long': '',
        'lat': '', 'alt': '', 'raty': 'a', 's': 's', 'm': 'm', 'igd': '', 'ibh': '', 'fp': '', 'oed': '', 'e': field,
        'res': 'y', 'resoc': '', 'tit': '', 'bu': '', 'ch': 'c', 'ce': 'f', 'js': 'f'
    }

    # Send the query to the MPC site
    data = urllib.parse.urlencode(query).encode('UTF-8')
    req = urllib.request.Request('https://cgi.minorplanetcenter.net/cgi-bin/mpeph2.cgi', data)
    response = urllib.request.urlopen(req)

    return response


def main():
    comet_name = 'P/2022 C1 (PANSTARRS)'
    mpc_orbit = get_comet_mpc_orbit(comet_name)
    print(mpc_orbit)

if __name__ == "__main__":
    main()
```

I think it should also be possible to send a list of comets to the query, since the ephmeris service accepts multiple comets each in a new line.

Clear skies

**Comment by brandon-rhodes** [comment]
> I'm also seeing warnings from mpc.py…

Thanks for sharing the warning, I hadn't heard about the Pandas plan for this breaking change. I've adjusted Skyfield's code and will try to get a release out soon!

Here's the commit I just made: 05760c6a2d6b3992726175a1b0c75c30a2bb7141

**Reply by ursomniac** [reply]
Thanks. Looking forward to the next release (there was something else tagged v1.42 that I wanted to try out, but I'm under-caffeinated at the moment and I can't remember what it was.

**Comment by ursomniac** [comment, marked as answer]
I found the problem.   The latest data files has commas in the last column in some rows: e.g.,  " 98, 1083".
Remove the commas from the data file and the code is happy.

Ideally the code should filter them out somehow when it goes to create the CSV to load into pandas.

**Reply by brandon-rhodes** [reply]
A fix will go into the next version of Skyfield, which I'll hopefully have the chance to release soon:

https://github.com/skyfielders/python-skyfield/issues/707
